### ISSUE TITLE
Introduce `ReportTable` component

### ISSUE THREAD

**Issue by felixarntz** [google/site-kit-wp#2249]
The `getDataTableFromData` function is overly complex and should be superseded by a `ReportTable` component.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* There should be a new functional component `ReportTable` in `assets/js/components/ReportTable.js`:
    * It should require `props.rows` as an array of objects of row data (in no specific structure as long as it is consistent across all rows)
    * It should require `props.columns` as an array of objects with the following properties:
        * `{string} title`: Column header title.
        * `{string} description`: Optional. Column header description, displayed as tooltip. Replaces today's `tooltip` property.
        * `{bool} primary`: Optional. Whether the column is primary, like today.
        * `{string} className`: Optional. Class name(s) to add to each `th` and `td` of the column
        * `{string} field`: Optional. The path to access the value to render for the column in each row.  
        If provided, the value rendered should be essentially `get( row, field )` via `lodash/get`  
        **Any further customization needed should use `Component` instead**
        * `{WPComponent} Component`: Optional. Component type to use for rendering cells for this column in each row.  
        If provided, `Component` will be rendered with the full `row` object passed as a prop
        **Takes precedence over `field`**
        * Each column must provide either a `field` or a `Component`
    * It should produce similar output like `getDataTableFromData`
    * It should render the header row using the title and description from each column in `columns`
    * It should render each row by iterating over each column definition for each row
        * If `column.Component`, it should render `<Component row={ row } />`
        * If `column.field` is provided and no `Component`, the value referred to by `column.field` should be rendered for that row
* The following components should be updated to rely on `ReportTable` instead of `getDataTableFromData`:
    * `DashboardPopularKeywordsWidget`
    * `DashboardPopularPagesWidget`
    * `DashboardTopEarningPagesWidget`
    * `AcquisitionSources`
    * `WPDashboardPopularPages` (Introduced in https://github.com/google/site-kit-wp/issues/2610)
* Visually, nothing should change about these four components using the new `ReportTable`.

## Implementation Brief
- Using `assets/js/components/`, create a new file `ReportTable.js` which will contain the `ReportTable` functional component with the following props:
	- `rows`: array of objects of row data (in no specific structure as long as it is consistent across all rows)
	- `columns`: array of objects with the following properties
		- `title`: {string} Column header title.
		- `description`: {string} Optional, column header description, displayed as tooltip. Replaces the current [tooltip property](https://github.com/google/site-kit-wp/blob/1680f18750ea7fc778d39a5a5feab64d0d1b24d0/assets/js/components/data-table.js#L137).
		- `primary`: {bool} Optional, whether the column is primary, [like today](https://github.com/google/site-kit-wp/blob/1680f18750ea7fc778d39a5a5feab64d0d1b24d0/assets/js/components/data-table.js#L134).
		- `field`: {string} Optional. The path to access the value to render for the column in each row. If provided, the value rendered should be essentially `get( row, field )` via `lodash/get`. Any further customization needed should use `Component` instead. See next point.
		- `Component`: {WPComponent} Optional. Component type to use for rendering cells for this column in each row. If provided, `Component` will be rendered with the full `row` object passed as a prop. Takes precedence over field.
	- `classNames`: {string|object} for adding additional class names to the table. For e.g `googlesitekit-table--with-list` which is used by many tables.
- The `ReportTable` component should output the same markup as what's [outputted](https://github.com/google/site-kit-wp/blob/1680f18750ea7fc778d39a5a5feab64d0d1b24d0/assets/js/components/data-table.js#L35) by `getDataTableFromData`. On top of the `googlesitekit-table__wrapper` class name, the `table` tag should have the following class name: `googlesitekit-table__wrapper--${ totalColumns }-col` where `totalColumns` is the length of columns. This is because we have some CSS which is applied when we have the [following class name](https://github.com/google/site-kit-wp/blob/1680f18750ea7fc778d39a5a5feab64d0d1b24d0/assets/sass/components/global/_googlesitekit-table.scss#L121): `googlesitekit-table__wrapper--2-col` whereby the primary column is 75% wide.
	```js
	<table className={ classnames(
		'googlesitekit-table__wrapper',
		`googlesitekit-table__wrapper--${ totalColumns }-col`,
	) }>
	```
	- In addition to the `googlesitekit-table` class name, add the following classnames:
		-  `googlesitekit-report-table` class name which will be used to target styles for the `ReportTable` component only and prevent any CSS conflicts with any `googlesitekit-table` styles.
		- `googlesitekit-table--with-list` to have appropriate styles for tables with list which is the case for all `ReportTable` components.
- To generate the table headers, loop through the `columns` prop and use the `title` and `description` values. If `primary` is set to true, add the `googlesitekit-table__head-item--primary` [class name to the](https://github.com/google/site-kit-wp/blob/1680f18750ea7fc778d39a5a5feab64d0d1b24d0/assets/js/components/data-table.js#L134) `th`. The `description` value is also used as the tooltip for the cell by passing it to the `data-tooltip` [attribute of the table header cell](https://github.com/google/site-kit-wp/blob/1680f18750ea7fc778d39a5a5feab64d0d1b24d0/assets/js/components/data-table.js#L137).
- To generate the table cells, loop through the array of `rows`,
	- check if there is a component defined for the column. If that is the case, then render the `Component`, passing the `row` as prop.
	```js
	if ( column.Component ) {
		<Component row={ row } />
	}
	```
	- if `Component` is not defined for the column, then use the value of the `field` property. We then use lodash's `get` function to retrieve the value. `get( row, column.field )` 
- Update the components below to use `ReportTable` instead of `getDataTableFromData`
	- `DashboardPopularKeywordsWidget`
	- `DashboardPopularPagesWidget`
	- `DashboardTopEarningPagesWidget`
	- `WPAnalyticsDashboardWidgetTopPagesTable`
	- `AnalyticsDashboardWidgetTopPagesTable`
	- `AnalyticsAdSenseDashboardWidgetTopPagesTable`
- Since the `ReportTable` component does not have the `options` prop which is present in the `getDataTableFromData` [function](https://github.com/google/site-kit-wp/blob/1680f18750ea7fc778d39a5a5feab64d0d1b24d0/assets/js/components/data-table.js#L44), all logic pertaining to the `options` prop should be handled outside of the `ReportTable` component. **Examples** below:
	- Using `DashboardPopularPagesWidget`,
		- The `useAdminURLs` is used in the `options` object. We can handle the logic like that:
			```js
			const adminURL = useSelect( ( select ) => select( CORE_SITE ).getAdminURL( 'googlesitekit-dashboard' ) );
			const siteURL = useSelect( ( select ) => select( CORE_SITE ).getReferenceSiteURL() );
			```
		- The columns are defined as follows:
			```js
			const columns = [
				{
					title: __( 'Most popular content', 'google-site-kit' ),
					primary: true,
					Component: ( { row } ) => {
						const [ title, url ] = row.dimensions;
						const link = url.startsWith( '/' ) ? url : '/' + url;

						return (
							<Fragment>
								<Link
									href={ addQueryArgs( adminURL, { permaLink: `${ siteURL }/${ link }` } ) }
									inherit
								>
									{ title }
								</Link>

								<Link
									className="googlesitekit-table__link--secondary"
									href={ link }
									inherit
									external
								>
									{ link }
								</Link>
							</Fragment>
						);
					},
				},
				{
					title: __( 'Views', 'google-site-kit' ),
					Component: ( { row } ) => numFmt( row.metrics[ 0 ].values[ 0 ], { style: 'decimal' } ),
				},
			];
			```
		- We can safely ignore the `googlesitekit-table__body-item-link` class name since there are no styles associated with it. For the  `googlesitekit-table__body-item-url` class name, we can rename the class name to `googlesitekit-table__link--secondary` and keep the existing styles in `assets/sass/components/global/_googlesitekit-table.scss`
		- Then include the `ReportTable` component, passing the raw rows array via the `rows` prop.
			```js
			<ReportTable
				rows={ data[ 0 ].data.rows }
				columns={ columns }
			/>
			```
	- `DashboardPopularKeywordsWidget`
		- The columns can be defined as:
			```js
			const columns = [
				{
					title: __( 'Keyword', 'google-site-kit' ),
					description: __( 'Most searched for keywords related to your content', 'google-site-kit' ),
					primary: true,
					Component: ( { row } ) => {
						const rowValue = row.keys[ 0 ];

						return (
							<Link
								className="googlesitekit-table__body-item-link"
								href={ addQueryArgs( baseServiceURL, { query: `!${ rowValue }` } ) }
								external
								inherit
							>
								{ rowValue }
							</Link>
						);
					},
				},
				{
					title: __( 'Clicks', 'google-site-kit' ),
					description: __( 'Number of times users clicked on your content in search results', 'google-site-kit' ),
					Component: ( { row } ) => numFmt( row.clicks, { style: 'decimal' } ),
				},
				{
					title: __( 'Impressions', 'google-site-kit' ),
					description: __( 'Counted each time your content appears in search results', 'google-site-kit' ),
					Component: ( { row } ) => numFmt( row.impressions, { style: 'decimal' } ),
				},
			];
			```
	- `DashboardTopEarningPagesWidget`
		- We are using the `cap` in the `options` object and this can be handle when building the rows for the `ReportTable` component as follows:
			```js
				<ReportTable
					rows={ data[ 0 ].data.rows.slice( 0, 5 ) }
					columns={ columns }
				/>
			```
	- `AnalyticsDashboardWidgetTopPagesTable`
		- `hideColumns` is used in the `options` object to hide certain cells on small screens. This can be done by adding the `hidden-on-mobile` class which will apply to all cells in the column.

### Test Coverage

* No changes expected.

### Visual Regression Changes

* No changes expected.

## QA Brief

* Enable `widgets.dashboard` and `widgets.pageDashboard` features
* Set up Site Kit with Analytics + AdSense active and with some amount of data available
* Verify widget report tables in these locations are visually consistent with their non-widget versions
	* Check column headers/sizing
	* number of rows
	* visible columns on mobile
* Site Kit main dashboard + page dashboards
	- ![image](https://user-images.githubusercontent.com/1621608/109629726-cd7a3000-7b4c-11eb-8d69-6d0c3f237995.png)
	- ![image](https://user-images.githubusercontent.com/1621608/109629792-e387f080-7b4c-11eb-8c1c-e88a3822a434.png)
	- ![image](https://user-images.githubusercontent.com/1621608/109630388-86d90580-7b4d-11eb-9072-bf4386fa1303.png)
	- `AcquisitionSources` was replaced by the new All Traffic Widget, so there is no longer a widget-based version of this
* WordPress Dashboard
	- ![image](https://user-images.githubusercontent.com/1621608/109630912-1bdbfe80-7b4e-11eb-8952-f74c53e62c37.png)
* Analytics Dashboard
	- ![image](https://user-images.githubusercontent.com/1621608/109631030-3f06ae00-7b4e-11eb-9187-b20223df1930.png)
* AdSense Dashboard
	- ![image](https://user-images.githubusercontent.com/1621608/109631242-6c535c00-7b4e-11eb-94f0-3d2107aceae9.png)

## Changelog entry

* Introduce `ReportTable` component for displaying reusable data tables.

**Comment by felixarntz** [comment]
@aaemnnosttv Can you review the ACs here and add your suggestions?

**Comment by aaemnnosttv** [comment]
@felixarntz I'll have more context on what to add here once the IB for #2228 is approved 🙂

**Comment by aaemnnosttv** [comment]
@felixarntz a few thoughts on the ACs here:

>  It should require props `rows` (array of objects or array of arrays) and `columns` (see below), and an optional `options`

* What if we provided `rows` as `children` instead?
* I'm not sure about keeping `options` as a prop - why not make each option its own prop?

> A new optional `urlField` should indicate the location where to get the URL or path from in each `row`

This seems redundant since every column will have a `field` property that maps to a value in the row

> `options` should support the same arguments that are supported in the legacy component

We should only support the options/props we need – there are several of legacy options that are no longer used or would be no longer relevant.

> A new optional `linkCallback` should be a function that receives a URL (found via `options.urlField`) and transforms it into a more specific URL (e.g. admin URL containing the input URL as query parameter). This argument is only relevant if `options.urlField` is also given, and it replaces `useAdminURLs` from the legacy component, which should not be supported by the new component.

As we found in the recent changes to `getDataTableFromData`, this required the use of a component type rather than a simple callback as originally planned. This is because each link may require selecting data from the data store to build the final URL (e.g. a report URL). I think using a component is the best way to support more complicated cell logic.

As such, I think it makes sense to allow for any column's cell to have a dedicated component to render it. This could be provided as `columns[ # ].Component` and be empty by default which would render `{ value }`. If `Component` is provided for the current column, as it is being iterated over for each row, then we should render the value using the given component instead:
```js
<Component row={ row } column={ column } primary={ primary }>{ value }</Component>
```
Something like that. The component could also have access to the full column and row data for maximum flexibility. This would keep the `ReportTable` simple as well by offloading special cases to the consumer.

That means that instead of `options.PrimaryLink` we could pass the component type as part of the definition of the column which would be more useful for other cells as well, not just links.

> If there is a `options.urlField`, the first column should also display that URL (combined with the site reference URL like today, if  the found URL starts with `/`, i.e. is a path).
>        * If there is a URL found from the above _and_ an `options.linkCallback` (function), that function should be called with the found URL, and the returned URL should be used for the _first_ link (i.e. the one that in the legacy function relies on the very specific `useAdminURLs` option).

This is a good example of such complexity that isn't needed in all cases and would be best extracted to a component to be passed through for rendering that column's cell.

---

Let me know what you think and we can get the ACs finalized here 👍

**Comment by felixarntz** [comment]
@aaemnnosttv 
> What if we provided `rows` as `children` instead?

What would be the benefit of that? My initial thoughts are that it makes more sense via a custom prop rather than as components. How would you envision them to be passed as children? A `ReportTableRow` component with the data for each row? We'd need to then also pass the `columns` definition to each of that which would be a bit tedious. I think passing the data as a custom prop to `ReportTable` makes that easier. Maybe worth calling it `dataRows` or so instead though?

> I'm not sure about keeping `options` as a prop - why not make each option its own prop?

Agreed, let's do that. Also, with your other suggestions most current options we probably won't need anyway.

I like your other ideas, particularly passing an optional `Component` for `columns` that require more specific display behavior. Let's do that. We could implement some common ones like `ColumnWithURL` that could be used like this:
```js
<ReportTable
    { ...props }
    columns=[
        {
            // Column that displays its value as a link to a specific URL.
            Component: ( { row, value } ) => <ColumnWithURL value={ value } url={ row.url } />,
        },
        {
            // Column that displays its value as a link to a specific URL and displays that URL below (linked too).
            Component: ( { row, value } ) => <ColumnWithURL value={ value } url={ row.url } displayURL={ row.url } />,
        },
        {
            // Column that displays its value as a link to a specific URL and displays another URL below (linked too).
            Component: ( { row, value } ) => <ColumnWithURL value={ value } url={ row.url } displayURL={ row.adminURL } />,
        },
        // ...
    ]
/>
```

I think we're on the same page here, can you update the ACs with your proposed approach and pass over to me for final look over?

**Comment by aaemnnosttv** [comment]
@felixarntz 

> What would be the benefit of that? My initial thoughts are that it makes more sense via a custom prop rather than as components. How would you envision them to be passed as children?

I was just thinking of `<ReportTable { ...props }>{ rows }</ReportTable>` since rows are like children of a table. It's kind of aligning the data array to being "the body" of the table. `children` doesn't have to be components (e.g. render props as children). Just an idea though, it would be okay as a `rows` or `dataRows` prop too I think.

> I like your other ideas, particularly passing an optional `Component` for `columns` that require more specific display behavior. Let's do that. We could implement some common ones like `ColumnWithURL` that could be used like this:

Thanks! I think that in particular really helps remove a lot of the complexity that's there now. As for the `row.url` – I don't think the `ReportTable` should be concerned with URLs at all. Instead, this could be how the rows are mapped before giving them to the table to render, or that could be handled by the `Component`.

E.g.
```js
<ReportTable
	rows={ rows }
	columns={
		{
			// Component that wraps the value in a link
			Component: ( row, value ) => <Link href={ get( row, 'dimensions.0' ) }>{ value }</Link>
		}
	}
/>
```

I agree that there may be value in having dedicated components for common patterns, perhaps like a `ColumnWithURL`. I think a lot can be done with the "anonymous wrapper" `Component` of the column alone though that we should keep `ReportTable` as simple as we can and not worry about extracting anything other than a value.

I'll keep this in mind when finalizing the ACs, and extract any obvious recurring patterns.

> I think we're on the same page here, can you update the ACs with your proposed approach and pass over to me for final look over?

👍

**Comment by aaemnnosttv** [comment]
@felixarntz I think this is ready for IB now. I've simplified it a bit so that it either uses `field` by default or uses the component provided (but the component would take precedence if present). I don't think it makes sense to require the `field` in all cases since that reference can be passed with the definition of the component too; I'm not sure it would make sense to use both at the same time. I think we kind of need to see how it will really look in use though, so we could leave that part a bit flexible to see what the IB looks like regarding that.

I've intentionally left out references to legacy bits like `useAdminURLs` and combining with the reference URL in some places to leave that to the IB. The `ReportTable` component itself should be very simple. It may make sense to extract some added utility components for doing this kind of thing but I don't think it needs explicit mentioning in the ACs.

Assigning back to you for a final look-over 👍

**Comment by felixarntz** [comment]
@aaemnnosttv New ACs look solid to me, moving over to IB.

**Comment by asvinb** [comment]
@aaemnnosttv @felixarntz in the AC, I noticed there is no mention of the `options` prop which is used in the `getDataTableFromData` [function](https://github.com/google/site-kit-wp/blob/1680f18750ea7fc778d39a5a5feab64d0d1b24d0/assets/js/components/data-table.js#L44). Is it intentional? Because there are a lot of things which depends on it such as whether we want to hide certain columns on mobile or whether we want to cap the data.
Another thing is related to this point in the AC: _Furthermore, the component should also be used by the refactored version of the WP dashboard widget (see #2075) and the refactored version of the admin bar menu (see #2076)._
I checked #2075 and as @tofumatt [commented](https://github.com/google/site-kit-wp/issues/2075#issuecomment-736780462), I don't see where we can use the `ReportTable` component since there are no `getDataTableFromData` calls. The same applies for #2076 . I don't see where we are using a table. 
Let me know!

**Comment by aaemnnosttv** [comment]
@asvin yes, that's intentional, no more `options` should be needed. The `ReportTable` component should be simple, but flexible.

Some the refactoring will require changes outside of where the function was used before; it won't be 100% handled by the new component.

> I checked #2075 and as @tofumatt [commented](https://github.com/google/site-kit-wp/issues/2075#issuecomment-736780462), I don't see where we can use the `ReportTable` component since there are no `getDataTableFromData` calls. 

It's used in `WPAnalyticsDashboardWidgetTopPagesTable`

https://github.com/google/site-kit-wp/blob/9ef87188198cc9efff1904a3d10e03c44fd847c4/assets/js/modules/analytics/components/wp-dashboard/WPAnalyticsDashboardWidgetTopPagesTable.js#L89

As for the Admin Bar menu, you're correct – we don't render any tables in the Admin Bar.

---

Regarding the `options` – there may be some left over here which are no longer used so I would double-check those before we see about how to support those with the new component.

Regarding `hideColumns` option, I think we just need to add support for a `className` in the definition of each column which would be added to the `th` and `td` for every cell in the column. I'll add this to the ACs.

Any other cases you feel are problematic?

**Comment by tofumatt** [comment]
IB looks good to me, but handing it over to @aaemnnosttv to quickly double-check it as it's a lengthy issue, he has more history on it, and he might have some particular comments/thoughts about the approach 🙂

**Comment by aaemnnosttv** [comment]
Thanks @tofumatt!

@asvin – You're on the right track, there's just a few things to clarify.

> `DashboardPopularPagesWidget` - The `useAdminURLs` is used in the `options` object. We can handle the logic like that:
> ```js
> const adminURL = useSelect( ( select ) => select( CORE_SITE ).getAdminURL( 'googlesitekit-dashboard' ) );
> const siteURL = useSelect( ( select ) => select( CORE_SITE ).getReferenceSiteURL() );
> ```

This doesn't quite describe how `options.useAdminURLs` would be reimplemented with the new component. Where are these select statements happening?
Maybe in this case it would make sense to introduce a `DashboardDetailsLink` component that accepts the `permaLink` as a prop and constructs the full URL internally before rendering a `Link` for it?

> `DashboardPopularKeywordsWidget` - The rows can be defined as:
> ```js
> const rows = data.map( ( row, i ) => {
> 	const query = row.keys[ 0 ];
> 	return [
> 		<Link
> 			className="googlesitekit-table__body-item-link"
> 			href={ addQueryArgs( baseServiceURL, { query: `!${ query }` } ) }
> 			external
> 			inherit
> 			key={ i }
> 		>
> 			{ query }
> 		</Link>,
> 		numFmt( row.clicks, { style: 'decimal' } ),
> 		numFmt( row.impressions, { style: 'decimal' } ),
> 	];
> } );
> ```

Regarding the mapping of rows, in some (maybe even most) cases this will no longer be necessary. In particular though, we shouldn't be writing JSX in a `data.map` callback – this is what `column.Component` is for.
In this case, the column for the query link would have a `Component` defined for it which would render the query link for every row. As a component, it can retrieve the value from its `row` prop and manipulate it as needed, use selectors, etc before rendering the `Link`. Similarly, even columns with a formatted value should use a `Component` as well, rather than pre-formatting the value in the data. This way the component has access to the raw value if needed.

E.g. the column for "Clicks" would have a column definition like so:

```js
{
	title: __( "Clicks" ),
	Component: ( { row } ) => numFmt( row.clicks, { style: 'decimal' } )
}
```

Note this is also easier to follow than it would be later in the previous post-mapped data where clicks would be something like `row[ 2 ]` since the mapped row returns an array.

If this is a common need, we could extract a `<Num>` component which might look something like this:
```js
// We can't define a `Number` component as it would conflict with the native type.
function Num ( { value, ...options } ) {
	return numFmt( value, options );
}
```
Just an idea, but I could see how something like this might be useful.

---

Regarding the classes that are currently in place **within the cell**, are these needed? E.g. `googlesitekit-table__body-item-link` or `googlesitekit-table__body-item-url`
Ideally the consumers of `ReportTable` shouldn't need to implement report-table-specific classes. Can we see about extracting some utilities or other components perhaps for decoupling the cell contents from the table?

> Finally include the `ReportTable` component, adding the `googlesitekit-table--with-list` class name to have the appropriate styling.

We shouldn't require all consumers to pass this class to the table, so if all `ReportTable`'s need it (which it seems like they do as the current option to negate it doesn't seem to be used), then we should include it in the base classes for the component.
If we need to keep it, we should make it flexible using a prop (e.g. `withList`), similar to what we do with `Widget`s and its `noPadding` prop. At first glance, it doesn't look like something that needs to be dynamic though.

> `AnalyticsDashboardWidgetTopPagesTable` - `hideColumns` is used in the `options` object to hide certain cells on small screens. This can be done by adding the `hidden-on-mobile` class name to specific cells when building columns and rows.

`column.className` would apply to all cells of the column. I think that's what you mean but just clarifying.

**Comment by asvinb** [comment]
@aaemnnosttv I think it'll be a bit overkilled to include the `DashboardDetailsLink` component since we are only using it in the `DashboardPopularPagesWidget` component. I think it's easier to maintain later down the road if we just define it in the `columns` for the table? What do you think? Also, for the question about reimplementing `options.useAdminURLs`, those are only examples and it's part of the `DashboardPopularPagesWidget` component.
Now I think I understand, trying to leverage the `Component` prop as much as possible, just passing the raw rows to the table. I've updated the IB to include that.
I've also updated the IB to remove any custom class names and added an alternative on how we can handle that.
Let me know your thoughts!

**Comment by aaemnnosttv** [comment]
@asvinb – IB is looking great, just a few points left to make:

1. There isn't a mention of using `column.description` for the tooltip (other than in the description of this option)
1.   
	> Since we are no longer using the `googlesitekit-table__body-item-link` and `googlesitekit-table__body-item-url` class names in `DashboardPopularPagesWidget`, edit `/assets/sass/components/global/_googlesitekit-table.scss` to add styles related to `googlesitekit-report-table`. We can safely ignore the `googlesitekit-table__body-item-link` class name since there are no styles associated with it. **To target the `googlesitekit-table__body-item-url` class name, we expect that it'll always be after any links. Thus we can style it as follows, which is the same styles as `googlesitekit-table__body-item-url`.**
	> ```scss
	> .googlesitekit-report-table {
	> 	a + a.googlesitekit-cta-link--external {
	> 		// styles
	> 	}
	> }
	> ```

	Rather than assuming this and applying this broadly using CSS, let's simply extract a class (or multiple) we can apply to those `Link`s that need it since we're in full control of the output of that cell.

Otherwise I think this is good to go 👍

**Comment by asvinb** [comment]
@aaemnnosttv IB updated to include more details about using `description` as tooltip for a table header cell. I also added a new class name for the links. Let me know if it's the correct assumption.

**Comment by aaemnnosttv** [comment]
@asvinb almost! 😄  Thanks for adding the `description` part. Only this part left to clarify:


What I meant was, rather than this (plus added CSS) `.googlesitekit-table__cell--multiple-links a + a.googlesitekit-cta-link--external`

```js
<div className="googlesitekit-table__cell--multiple-links">
	<Link
		href={ addQueryArgs( adminURL, { permaLink: `${ siteURL }/${ link }` } ) }
		inherit
	>
		{ title }
	</Link>

	<Link
		href={ link }
		inherit
		external
	>
		{ link }
	</Link>
</div>
```

Can't we do this?

```js
<div>
	<Link
		href={ addQueryArgs( adminURL, { permaLink: `${ siteURL }/${ link }` } ) }
		inherit
	>
		{ title }
	</Link>

	<Link
		className={ /* custom link styles here */ }
		href={ link }
		inherit
		external
	>
		{ link }
	</Link>
</div>
```

Maybe we can even continue using `googlesitekit-table__body-item-url` or rename it to be something like `googlesitekit-table-item-url` ?

**Comment by asvinb** [comment]
@aaemnnosttv Now I see what you mean. I wanted initially to avoid using custom class names but I guess we can make an exception for this one. I would recommend we rename the class name to `googlesitekit-table__link--secondary` though since it's more in line with how it'll be used. The IB has been updated to reflect that.

**Comment by aaemnnosttv** [comment]
Great thanks @asvinb 👍 

IB ✅

**Comment by wpdarren** [comment]
### QA Update: ✅ Pass

**Verified:**

Crossed checked between `widgets.dashboard` and `widgets.pageDashboard` features enabled and disabled. 

- Verified widget report tables are visually consistent with their non-widget versions, including column headers/sizing
and number of rows. 
- Verified columns on mobile.
